source/agora/consensus/validation/Block.d

```diff
@@ -477,24 +481,43 @@ public bool isGenesisBlockValid (const ref Block genesis_block)
     return isGenesisBlockInvalidReason(genesis_block) is null;
 }
 
-/// Ditto but returns `bool`, only usable in unittests
+/// Ditto but returns `bool` and logs reason if fails, only usable in unittests
 version (unittest)
-{
 public bool isValid (const ref Block block, Height prev_height,
+    Hash prev_hash, scope UTXOFinder findUTXO,
+    size_t active_enrollments, scope PayloadChecker checkPayload,
+    scope EnrollmentFinder findEnrollment, ulong enrollment_cycle = 0,
+    string file = __FILE__, size_t line = __LINE__) nothrow @safe
+{
+    string reason = isInvalidReason(block, prev_height, prev_hash, findUTXO,
+        active_enrollments, checkPayload, findEnrollment);
+    if (reason !is null)
+        log.trace("[{}:{}] block height {} is invalid: {}",
+            file, line, block.header.height, reason);
+    return reason is null;
+}
+
+/// Ditto but returns `bool`, only usable in unittests
+version (unittest)
+public bool isNotValid (const ref Block block, Height prev_height,
     Hash prev_hash, scope UTXOFinder findUTXO,
     size_t active_enrollments, scope PayloadChecker checkPayload,
     scope EnrollmentFinder findEnrollment) nothrow @safe
 {
-    return isInvalidReason(block, prev_height, prev_hash, findUTXO,
-        active_enrollments, checkPayload, findEnrollment) is null;
+    string reason = isInvalidReason(block, prev_height, prev_hash, findUTXO,
+        active_enrollments, checkPayload, findEnrollment);
```

Comment:
Could you add a log call here too?

Reply:
But then it will log the cases where we expect it to fail which is why I create seperate functions for the validation. IsValid will log if unexpected failure and isNotValid will not log as it is the failure we are testing for.

Reply:
Oh right I misunderstood. It only makes sense to log in one case.